pysid/identification/recursive.py

```diff
@@ -0,0 +1,266 @@
+
+    Returns
+    -------
+    bool.
+
+    """
+    #Caso o erro tenha variado menos que th entre as iterações, retorna False e sai do while
+    if error[0,0] != 0 :
+        for i in range(error.shape[1]):
+            if (abs((error[0,i] - error[1,i])/error[0,i])) > th :
+                return True
+        return False
+    else:
+        return True
+    
+def emq(na,nb,nc,nk,u,y,th,n_max):
+    """
+    
+    Performs the extended least squres algorithm on u,y data,
+    indentifing A,B and C polynomials with na,nb and nc degree respectively.
+    
+    **The version takes presuposses that the order of every polynomial (in not siso cases)
+    is the same. So A11 has the same order as any other Aij , same goes for B11 with any Bij
+    
+    Parameters
+    ----------
+    na : int
+        Degree of A polynomial
+    nb : int
+        Degree of B polynomial
+    nc : int
+        Degree of C polynomial
+    nk : int
+        minimum delay for B polynomial
+    u : numpy array
+        Array (or array of arrays) contaning the inputs chronologically
+    y : numpy array
+        Array (or array of arrays) contaning the outputs chronologically
+    th : float
+        Treshold
+        Minimum difference between the quadratic sum of two consecutive errors needed to assume that there was convergence
+    n_max : int
+        Maximum number of iterations allowed
+        
+    Returns
+    -------
+    m : pysid polymodel
+
+    """
+    nu = u.shape[1]
+    ny = y.shape[1]
+    
+    nbk = nb+nk
+    
+    w = max(na,nbk,nc)
+    # psi = zeros([y.shape[0]-w,nu*(nb+1)+ny*na])
+    
+    y_sol = y[w:] #saida usada para theta resoluçaõ do problema linear
+    u_reg = zeros([u.shape[0]-w,nu*(nb+1)])
+    cont = 0
+    # nb quantos atrasos vou olhar
+    # nk onde eu começo a olhar(-1, -2, -3,...)
+    for c in range(nu): #para cada entrada
+        for i in range((nb+1)): #u no regressor vai de u[k-nk] até u[k-nb-nk+1]
+            u_reg[:,cont] = u[w-nbk+i:u.shape[0]-nbk+i,c]
+            cont = cont + 1
+    cont = 0
+    y_reg = zeros([y.shape[0]-w,ny*na])
+    for c in range(ny):
+        for i in range(na):
+            y_reg[:,cont] = -y[w-na+i:y.shape[0]-na+i,c]
+            cont = cont + 1 
+    psi = concatenate((u_reg,y_reg),axis=1)
+    
+    #MQ padrão : theta = pseudo_inversa(psi) * y
+    theta_mq = qrsolm(psi,y_sol)
+    
+    # montar psi_emq (1° vez)
+    
+    psi_emq = zeros((y_sol.shape[0]-nc,psi.shape[1]+nc,ny))
+    res = y_sol - matmul(psi,theta_mq)
+    #Agora cada saida tem seu próprio regressor, já que os residuos são diferentes
+    for i in range(ny):
+        psi_emq[:,:psi.shape[1],i] = psi[nc:,:]
+        for j in range(nc):
+            res_i = res[j:-nc+j,i]
+            psi_emq[:,psi_emq.shape[1]-nc+j,i] = res_i
+    
+    
+    #Primeiro theta_emq
+    theta_emq = zeros((psi_emq.shape[1],ny))
+    
+    #é preciso pegar o theta certo, o que foi calculado com o residuo da saida correspondente
```

Comment:
It is better to use only English for all comments, specially for those who do not use UTF-8 as coding. "é" it is usually not printed.